google-startup-scripts/usr/share/google/onboot

```diff
@@ -112,6 +112,41 @@ function print_ssh_key_fingerprints() {
   return 0
 }
 
+function check_for_connection() {
+  local count=0
+  local return_code=0
+  local url="http://metadata/computeMetadata/v1/instance/?recursive=True"
+
+  log "Checking for metadata server connection."
+  while true; do
+    ((count++))
+    curl -f ${url} -H "Metadata-Flavor: Google" 2>/dev/null
+    return_code=$?
+    case $return_code in
+      # No error.  Connection is active.
+      0) break;;
+      # Failed to resolve host or connect to host.  Retry indefinitely.
+      6|7) sleep 1.0
+        log "Waiting for metadata server, attempt ${count}"
+        # After 7 minutes, add a console message denoting a probable network
+        # issue. On systems using dhclient there is an attempt to obtain an IP
+        # for 60 seconds followed by a 5 minute wait period. After 7 minutes,
+        # this cycle will have run through twice. After this period of time, it
+        # is not known when a DHCP lease might be obtained and the network
+        # interface fully operational.
+        if ((count >= 421)); then
+          log "There is likely a problem with the network."
+        fi
+        continue;;
+      # A genuine error but a connection exists.
+      *) log "Extraneous error getting metadata ${return_code}"; break;;
+    esac
+  done
+  # Return the last return code we got.
+  return ${return_code}
+}
+
+check_for_connection
```

Comment:
This should probably happen after set_interrupts and virtionet_irq_affinity, which are performance tweaks for disk and network and which don't hit the network at all (only in-guest sources like /sys and /proc). I agree this should precede first_boot since that hits the metadata server.